x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -0,0 +1,740 @@
+            JwtRealmSettings.HEADER_CLIENT_AUTHORIZATION,
+            JwtRealmSettings.SUPPORTED_CLIENT_AUTHORIZATION_TYPE_SHARED_SECRET,
+            true
+        );
+
+        return new JwtAuthenticationToken(authorizationParameterValue, clientAuthorizationSharedSecretValue);
+    }
+
+    @Override
+    public void authenticate(final AuthenticationToken authenticationToken, final ActionListener<AuthenticationResult<User>> listener) {
+        this.ensureExpectedValueForInitialized(true);
+        if (authenticationToken instanceof JwtAuthenticationToken jwtAuthenticationToken) {
+            LOGGER.trace("Realm [{}] supports JwtAuthenticationToken.", super.name());
+            final JWSHeader jwsHeader = jwtAuthenticationToken.getJwsHeader();
+            final JWTClaimsSet jwtClaimsSet = jwtAuthenticationToken.getJwtClaimsSet();
+            final String clientAuthorizationSharedSecret = jwtAuthenticationToken.getClientAuthorizationSharedSecret().toString();
+
+            // Filter steps (before any validation)
+
+            // 1. Skip JWT if signature algorithm does not match any of the signature algorithms allowed by this realm.
+            final JWSAlgorithm jwsSignatureAlgorithm = jwsHeader.getAlgorithm();
+            if ((jwsSignatureAlgorithm == null) || (this.allowedSignatureAlgorithms.contains(jwsSignatureAlgorithm.getName()) == false)) {
+                final String msg = String.format(
+                    Locale.ROOT,
+                    "Realm [%s] does not allow signature algorithm [%s]. Allowed signature algorithms are %s.",
+                    super.name(),
+                    jwsSignatureAlgorithm,
+                    this.allowedSignatureAlgorithms
+                );
+                LOGGER.debug(msg);
+                listener.onResponse(AuthenticationResult.unsuccessful(msg, null));
+                return;
+            }
+            LOGGER.debug(
+                "Realm [%s] allows signature algorithm [%s]. Allowed signature algorithms are %s.",
+                super.name(),
+                jwsSignatureAlgorithm,
+                this.allowedSignatureAlgorithms
+            );
+
+            // 2. Skip JWT if issuer does not match the issuer allowed by this realm.
+            final String jwtIssuer = jwtClaimsSet.getIssuer();
+            if ((jwtIssuer == null) || (this.allowedIssuer.contains(jwtIssuer) == false)) {
+                final String msg = String.format(
+                    Locale.ROOT,
+                    "Realm [%s] does not allow issuer [%s]. Allowed issuer is [%s].",
+                    super.name(),
+                    jwtIssuer,
+                    this.allowedIssuer
+                );
+                LOGGER.debug(msg);
+                listener.onResponse(AuthenticationResult.unsuccessful(msg, null));
+                return;
+            }
+            LOGGER.debug("Realm [%s] allows issuer [%s]. Allowed issuer is [%s].", super.name(), jwtIssuer, this.allowedIssuer);
```

Comment:
```suggestion
            LOGGER.trace("Realm [%s] allows issuer [%s]. Allowed issuer is [%s].", super.name(), jwtIssuer, this.allowedIssuer);
```

Reply:
I prefer debug logs to show the customer why some JWT realms are skipped before the authenticating JWT realm is found to attempt JWT validation.

Reply:
Done